### PR TITLE
Add OMCoreComplexRTLModule and support for om signals

### PR DESCRIPTION
**Pull request by derekpappas**
**Related issue**: 

**Type of change**: feature request

**Impact**: API addition (no impact on existing code)

**Development Phase**: implementation

**Release Notes**

Add OMCoreComplexRTLModule and support for om signals

### CODE REVIEW

File: src/main/scala/diplomaticobjectmodel/model/OMCoreComplex.scala [hunk 1]
```diff
@@ -2,8 +2,27 @@
...
 package freechips.rocketchip.diplomaticobjectmodel.model
 
+case class CoreComplexRTLInterface(
+  clocks: List[OMClock] = Nil,
+  clockRelationships: List[OMClockRelationship] = Nil,
+  resets: List[OMRTLReset] = Nil,
+  statuses: List[OMStatus] = Nil,
+  localInterrupts: Option[OMInterruptSignal] = None,  // E.g. local_interrupts_X (all local interrupts for core X)
+  globalInterrupts: Option[OMInterruptSignal] = None,  // E.g. global_interrupts
+  machineExternalInterrupts: Option[OMInterruptSignal] = None,  // E.g. meip_X
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/model/OMCoreComplex.scala]: These should be Seqs, not Options. I believe the current implementation will have one instance of the local and the machine external interrupts per core. The global interrupts might be a singleton.

File: src/main/scala/diplomaticobjectmodel/HasLogicalTreeNode.scala [hunk 2]
```diff
@@ -4,6 +4,7 @@ package freechips.rocketchip.diplomaticobjectmodel
 
 
 import freechips.rocketchip.diplomaticobjectmodel.logicaltree._
+import freechips.rocketchip.diplomaticobjectmodel.model.OMInterruptSignal
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/HasLogicalTreeNode.scala]: Why is this being imported here?

File: src/main/scala/diplomaticobjectmodel/model/OMRTLModule.scala [hunk 3]
```diff
@@ -31,22 +35,79 @@ trait Asynchronous extends Synchronicity
...
+  synchronicity: Option[Synchronicity],
+  uid: Int,
+  _types: Seq[String] = Seq("OMReset", "OMSignal", "OMRTLComponent", "OMCompoundType")
+) extends OMSignal
+
+// Status
+trait OMStatus extends OMSignal {
+  def description: Option[String]
+  def name: String
+  uid: Int
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/model/OMRTLModule.scala]: I think you mixed the syntax for case classes with traits here.

File: src/main/scala/diplomaticobjectmodel/model/OMRTLModule.scala [hunk 4]
```diff
@@ -31,22 +35,79 @@ trait Asynchronous extends Synchronicity
...
+}
+
+trait OMCoreStatus extends OMStatus {
+  def description: Option[String]
+  def name: String
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/model/OMRTLModule.scala]: These two fields are already defined in `OMStatus`, so you don't need to repeat them here.

File: src/main/scala/subsystem/BaseSubsystem.scala [hunk 5]
```diff
@@ -74,7 +74,27 @@ abstract class BaseSubsystem(implicit p: Parameters) extends BareSubsystem with
...
+    val rtlIfc = CoreComplexRTLInterface(
+      clocks = Nil,
+      clockRelationships = Nil,
+      resets = Nil,
+      statuses = Nil,
+      localInterrupts = None, // E.g. local_interrupts_X (all local interrupts for core X)
+      globalInterrupts = None, // E.g. global_interrupts
+      machineExternalInterrupts = None, // E.g. meip_X
+      testModeSignals = Nil // E.g. debug_psd_test_mode and debug_psd_test_mode_reset
+    )
```
Comment (richardxia) [inline src/main/scala/subsystem/BaseSubsystem.scala]: What is the purpose of filling in all of these fields with `None` and `Nil`? If the whole function returns an `Option[OMCoreComplexRTLModule]`, why not just return `None` for the whole thing instead of filling out null values for each individual field?

File: src/main/scala/system/ExampleRocketSystem.scala [hunk 6]
```diff
@@ -37,6 +37,8 @@ class ExampleRocketSystem(implicit p: Parameters) extends RocketSubsystem
     sbus.coupleTo("coherence_manager") { in :*= _ }
     mbus.coupleFrom("coherence_manager") { _ :=* BankBinder(mbus.blockBytes * (nBanks-1)) :*= out }
   }
+
+  override def getOMGlobalInterruptSignals() = None
```
Comment (hcook) [inline src/main/scala/system/ExampleRocketSystem.scala]: There actually are interrupt signals, in trait `HasAsyncExtInterrupts` which is mixed in here.

File: src/main/scala/diplomaticobjectmodel/model/OMRTLModule.scala (hunk 3 above)
Comment (mwachs5) [reply src/main/scala/diplomaticobjectmodel/model/OMRTLModule.scala]: Not sure why you are calling this out here in particular... there is already quite a mix of those in this file.

File: src/main/scala/subsystem/BaseSubsystem.scala (hunk 5 above)
Comment (mwachs5) [reply src/main/scala/subsystem/BaseSubsystem.scala]: I'm assuming this is where designer should fill in their contents?

File: src/main/scala/diplomaticobjectmodel/model/OMRTLModule.scala (hunk 3 above)
Comment (richardxia) [reply src/main/scala/diplomaticobjectmodel/model/OMRTLModule.scala]: No, I'm specifically referring to the fact that this particular trait uses both `def` and non-`def`. In every other thing in this file, all the Traits only use `def` and all the case classes just use constructor arguments. I don't actually understand what this syntax means, and I originally thought it would be a syntax error. Can you explain what this syntax means, since I haven't seen it before? Traits in Scala aren't allowed to define a constructor, so I don't really know what this specific `uid` line means.

File: src/main/scala/subsystem/BaseSubsystem.scala (hunk 5 above)
Comment (richardxia) [reply src/main/scala/subsystem/BaseSubsystem.scala]: But there's no purpose to this object instance; it doesn't represent anything that couldn't be represented with a `None`. If someone else wanted to later fill in a value, they could easily just pattern match on the `None` and fill in their own values in the same way.

Comment (mwachs5) [reply src/main/scala/subsystem/BaseSubsystem.scala]: Ok, I think I understand what you're saying... let me rephrase it here and you sanity check?

It's basically expected that someone is going to override this `  def getCoreComplexRTLModule(): Option[OMCoreComplexRTLModule] = ` entirely. And therefore, there is no reason that this default version shouldn't just return `None`, vs bothering to fill out a bunch of useless instances.

is that what you mean?

Comment (richardxia) [reply src/main/scala/subsystem/BaseSubsystem.scala]: Yes, that's what I mean, although I don't know exactly what the original intent of this function was.

Comment (mwachs5) [reply src/main/scala/subsystem/BaseSubsystem.scala]: I assumed that Derek was just creating the skeleton, then would ask designer to come modify *this code* to fill in the appropriate values, as part of this PR. At any rate I will attempt to do *that* to see if it makes sense, otherwise I'll just return `None` as you explained.
